sdk/storage/azure-storage-datamovement/inc/azure/storage/datamovement/job_engine.hpp

```diff
@@ -0,0 +1,233 @@
+#include "azure/storage/datamovement/transfer_engine.hpp"
+#include "azure/storage/datamovement/utilities.hpp"
+
+namespace Azure { namespace Storage {
+  namespace _detail {
+    struct JobPlan;
+    class JobEngine;
+  } // namespace _detail
+  namespace _internal {
+
+    class TransferEnd final {
+    public:
+      enum class EndType
+      {
+        Uninitialized,
+        LocalFile,
+        LocalDirectory,
+        AzureBlob,
+        AzureBlobFolder,
+      };
+
+      std::string ToString() const;
+      static TransferEnd FromString(const std::string& str, const TransferCredential& credential);
+      static TransferEnd CreateFromLocalFile(const std::string& path);
+      static TransferEnd CreateFromLocalDirectory(const std::string& path);
+      static TransferEnd CreateFromAzureBlob(Blobs::BlobClient blobClient);
+      static TransferEnd CreateFromAzureBlobFolder(Blobs::BlobFolder blobFolder);
+
+    private:
+      EndType m_type = EndType::Uninitialized;
+      std::string m_url;
+      Nullable<Blobs::BlobClient> m_blobClient;
+      Nullable<Blobs::BlobFolder> m_blobFolder;
+
+      friend struct _detail::JobPlan;
+      friend class _detail::JobEngine;
+    };
+
+    struct JobModel final
+    {
+      TransferEnd Source;
+      TransferEnd Destination;
+    };
+
+    struct HydrationParameters final
+    {
+      TransferCredential SourceCredential;
+      TransferCredential DestinationCredential;
+      std::function<void(TransferProgress&)> ProgressHandler;
+      std::function<void(TransferError&)> ErrorHandler;
+    };
+  } // namespace _internal
+
+  namespace _detail {
+    struct TaskModel final
+    {
+      int32_t NumSubtasks = 0;
+      std::string Source;
+      std::string Destination;
+      int64_t ObjectSize = -1;
+      int64_t ChunkSize = -1;
+      std::map<std::string, std::string> ExtendedAttributes;
+
+      std::string ToString() const;
+      static TaskModel FromString(const std::string& str);
+    };
+
+    struct PartGenerator final
+    {
+      std::string Source;
+      std::string Destination;
+      std::string ContinuationToken;
+
+      std::string ToString() const;
+      static PartGenerator FromString(const std::string& str);
+    };
+
+    struct JobPart final
+    {
+      JobPart() = default;
+      JobPart(JobPart&&) = default;
+      ~JobPart();
+
+      static std::pair<JobPart, std::vector<TaskModel>> LoadTasks(JobPlan* plan, uint32_t id);
+
+      static void CreateJobPart(
+          uint32_t id,
+          const std::string& jobPlanDir,
+          const std::vector<TaskModel>& tasks);
+
+      _internal::MovablePtr<JobPlan> m_jobPlan;
+      uint32_t m_id = 0;
+      size_t m_numDoneBits = 0;
+      std::unique_ptr<std::atomic<size_t>> m_numUndoneBits
+          = std::make_unique<std::atomic<size_t>>(0);
+      std::unique_ptr<_internal::MemoryMap> m_mappedFile;
+      bool* m_doneBitmap = nullptr;
+    };
+
+    struct JobPlan final
```

Comment:
Add some comments to brief its functionality and logic flow.

Reply:
Added